File: 0125-valid-palindrome/0125-valid-palindrome.cpp
class Solution {
public:
    /*bool isPalindrome(string s,int i=0) {
       int n=s.size();
        if(i>n/2) return true;
        if(s[i]!=s[n-i-1]) return false;
        return isPalindrome(s,i+1);
        
    }*/
    bool isPalindrome(string s){
        int l=0,r=s.size()-1;
        while(l<r){
            while(l<r && !isalnum(s[l])){
                l++;
            }
            while(l<r && !isalnum(s[r])){
                r--;
            }
            if(tolower(s[l++])!=tolower(s[r--])){
                return false;
            }
        }
        return true;

    }
    
};

File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        right = len(s) - 1
        left = 0

        while(left < right):
            if s[left].isalnum() and s[right].isalnum():
                if s[left].lower() != s[right].lower():
                    return False
                else:
                    left += 1
                    right -= 1
            else:
                if not s[left].isalnum():
                    left += 1
                else:
                    right -= 1

        return  True

# TC - O(N)
# SC - O(1)
